### ISSUE TITLE
getDefaultMaterialProperty does not register the dependency resolution!

### ISSUE THREAD

**Issue by dschwen** [idaholab/moose#6063]
In the plug and play TensorMechanics system the order of material blocks in the input file is important. In particular the strain calculation material has to appear _after_ the eigenstrain calculation material. This is because the strain class obtains an eigenstrain material property using the `getDefaultMaterialProperty` call, which - unlike the `getMaterialProperty` call - does _not_ enforce proper dependency resolution between material classes. 

With the wrong input file order the strain class will get executed before the eigenstrain class thus using an outdated eigenstrain value from the previous element (which might be on the opposite end of the mesh and have a completely different eigenstrain state!).

FYI @tonkmr

**Comment by friedmud** [comment]
Yikes! Good catch!

Derek

**Comment by mangerij** [comment]
How did you figure that out?

**Comment by dschwen** [comment]
Thanks to @laagesen's diligent verification work we encountered this problem. We spent about 2.5h in front of his computer reworking the input file in every possible way (initial conditions, boundary conditions, change of axis etc.), we looked at all material properties (which looked fine, as the output system had access to the properties at the right time).
We observed something that looked like a periodicity effect in just one dimension (change of axis showed qualitatively different behavior). Eventually we reordered the materials and I started investigating if the construction order mattered. Looking into the source code we narrowed it down to the `getDefaultMaterialProperty` call. Then the pseudo periodicity in the x direction made sense (a 1 element lag in the element loop).

It was a process.
